### PR TITLE
EZP-24329: prototype multicore Solr search engine

### PR DESCRIPTION
**Pull request by pspanja**
This PR resolves https://jira.ez.no/browse/EZP-24329

This implements a multicore prototype of Solr search engine indexing single language per core.

That means the same Content can be indexed multiple times, in one core per Content's translation. Depending on the passed `$fieldFilters` parameter, same query will be distributed towards multiple cores. Since each indexed document contains only single translation, `$languageCode` parameter of the Field sort clause is not used. Note that it also means all conditions necessarily apply on a **single** Content translation. No grouping is used since the same Content will match only once in all queried cores. Logic that takes care of that is following:

```php
array(
    "languages" => array(
        "cro-HR",
        "eng-GB",
        "ger-DE",
    ),
    "useAlwaysAvailable" => true,
)
```

From the `$fieldFilters` parameter described above, following filtering condition will be appended to the query:

```
indexed_language:cro-HR 
OR (indexed_language:eng-GB AND NOT content_translations:cro-HR) 
OR (indexed_language:ger-DE AND NOT content_translations:eng-GB AND NOT content_translations:cro-HR)
OR indexed_language_is_main_and_always_available:true
```

It is obvious that this solution will create progressively more complex condition for each additional language used. That is resolved as: language fallback should be used with measure.

Also, it shows that `$fieldFilters` parameter is here interpreted as language configuration, containing prioritized list of languages and always available switch.

Additionaly, if always available fallback is used, all cores will need to be queried, since we can't know which one contains a Content in a main language that is also always available.

Previous implementation of field filtering (implemented for Legacy and Elasticsearch) engines was working with a complete translation data in the same index. Tests implemented for that approach are not applicable for this one. For that reason field filter tests were copied and adapted, but that is really a temporary solution and dedicated set of test will need to be implemented if this is accepted.

#### Open questions

1. Since always available fallback is pretty much the default case, is distributing the search query to all cores acceptable? It obviously introduces more resource requirements.
2. Since `$fieldFilter` parameter is here interpreted as language configuration (prioritized language list + AA fallback), does it make sense to adapt the service to that as well? 

    That would basically mean loading fields only for matched translation, the one that will end up being displayed on a siteaccess. It obviously moves that control from MVC to Public API.

#### Depends on

1. https://github.com/ezsystems/ezpublish-kernel/pull/1257

### CODE REVIEW

**Comment by pspanja** [comment]
Expecting last commit to pass (aside from unit tests, which depend on https://github.com/ezsystems/ezpublish-kernel/pull/1257).
this is now ready for review, ping @andrerom @bdunogier @lolautruche @dpobel

File: eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ */
+
+namespace eZ\Publish\Core\Search\Solr\Content\Gateway;
+
+use eZ\Publish\SPI\Search\FieldType;
+
+/**
+ *
+ */
+interface EndpointProvider
```
Comment (andrerom) [inline eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider.php]: phpdoc, explaining what this is used for

File: eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider/TranslationEndpointProvider.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+ * @version //autogentag//
+ */
+
+namespace eZ\Publish\Core\Search\Solr\Content\Gateway\EndpointProvider;
+
+use eZ\Publish\Core\Search\Solr\Content\Gateway\EndpointProvider;
+use RuntimeException;
+
+/**
+ *
```
Comment (andrerom) [inline eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider/TranslationEndpointProvider.php]: phpdoc

File: eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ * @version //autogentag//
+ */
+
+namespace eZ\Publish\Core\Search\Solr\Content\Gateway;
+
+use eZ\Publish\SPI\Search\FieldType;
+
+/**
+ *
```
Comment (andrerom) [inline eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider.php]: phpdoc, explaining what this is used for

File: eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider/TranslationEndpointProvider.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+{
+    /**
+     * @var array
+     */
+    protected $endpointMap;
+
+    public function __construct()
+    {
+        $this->endpointMap = array(
+            "eng-GB" => "http://localhost:8983/solr/core0",
```
Comment (andrerom) [inline eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider/TranslationEndpointProvider.php]: could we have a mapping that consists of repository(or db) name + language name or similar? (to make sure multi repository is possible)

**Comment by andrerom** [comment]
some review notes:
- not possible to do delete request in one go with parms for cores to affect?
- endpoints seems rather hard coded, what is the plan for this?

> $languageCode parameter of the Field sort clause is not used

Ok, was thinking this param should be removed anyway to simplify the sort clause for the sql impl, is this the direction your are thinking of to?

> It is obvious that this solution will create progressively more complex condition for each additional language used. That is resolved as: language fallback should be used with measure.

True, any feedback on additional queries to use for language core selection @paulborgermans ?

File: eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider/TranslationEndpointProvider.php (hunk 4 above)
Comment (pspanja) [inline eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider/TranslationEndpointProvider.php]: Yes, but that will be resolved by semantic configuration, and the provider service won't know about it.

**Comment by pspanja** [comment]
> not possible to do delete request in one go with parms for cores to affect?

Correct, it needs to be sent to each core separately. Solr Cloud might provide something for that though.

> endpoints seems rather hard coded, what is the plan for this?

My plan is to take care of semantic configuration through separate issue.

> Ok, was thinking this param should be removed anyway to simplify the sort clause for the sql impl, is this the direction your are thinking of to? (as far as I understood it, this is the param responsible for the filtering behavior right?)

Not really, it is about the possibility to mach different Types of Content, so the sort field might not always be available. For that reason we would still need to `LEFT JOIN`. Some simplification might be possible, but it goes together with field filters change, and I need to look deeper into it.

**Comment by andrerom** [comment]
> Not really, it is about the possibility to mach different Types of Content

you mean different kind of content types?
Maybe we need to introduce restrictions that it needs to be on a certain type, across several types will result in strange results.

**Comment by pspanja** [comment]
> you mean different kind of field types?
Maybe we need to introduce restrictions that it needs to be on a certain type, across several types will result in strange results.

Sort clause targets a FieldDefinition in a ContentType, so the intended FieldType is known. If multiple ContentTypes are returned by the filter, sort value for those not targeted by the sort clause will be `null` and those will be always sorted last, for both ascending and descending order. This is how Solr and Elasticsearch work out of the box, and Legacy search engine was made to work in the same way.

File: eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider/TranslationEndpointProvider.php (hunk 2 above)
Comment (pspanja) [inline eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider/TranslationEndpointProvider.php]: Added in https://github.com/ezsystems/ezpublish-kernel/commit/7a3e17cd0211a48bc65a9b328e2d2e6b951a2e78

File: eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider.php (hunk 3 above)
Comment (pspanja) [inline eZ/Publish/Core/Search/Solr/Content/Gateway/EndpointProvider.php]: Added in https://github.com/ezsystems/ezpublish-kernel/commit/7a3e17cd0211a48bc65a9b328e2d2e6b951a2e78

**Comment by andrerom** [comment]
+1

**Comment by kore** [comment]
+1

**Comment by pspanja** [comment]
Manually merged in https://github.com/ezsystems/ezpublish-kernel/commit/8e50f0a0dab002dfce573dfd304087af6cdcf82c.

**Pull request closed without merging**